### DISCUSSION TITLE
[Feature Request] Programmatic access to a target's expanded inputs

### DISCUSSION THREAD

**Discussion by MJez29** [nrwl/nx#23344] [Feature Requests]
- [x] I'd be willing to implement this feature ([contributing guide](https://github.com/nrwl/nx/blob/master/CONTRIBUTING.md))

## Description

I would like a way to programmatically access the expanded inputs for a target

## Motivation

<img width="1911" alt="Screenshot 2024-01-23 at 3 45 29 PM" src="https://github.com/nrwl/nx/assets/20645805/b79803b6-3a76-4424-a0e6-840017d16975">

The input list is currently viewable in Nx Console when clicking on a target in the task graph which can be useful for debugging a specific task but I believe the real power of this feature comes with being able to write tooling to verify that the files listed as inputs do not import files that are not listed as inputs.

In order to make that happen there needs to be a way to get the expanded inputs from Nx for a target

## Suggested Implementation

This could either be accessible via devkit

```ts
function getExpandedTargetInputs(projectGraph: ProjectGraph): string[]
```

or the cli

```bash
nx show inputs [target] [project]
```

**Comment by FrozenPandaz** [comment]
I've converted this to a discussion so we can discuss more easily.

**Comment by studds** [comment]
I would use this.

The situation is this: I want to answer the question "which products will be changed by this PR". The answer to this question flows into which e2e tests we require to merge the PR (as some of our integration tests take a long time to run), and to post an informative message of "these products are impacted by this PR. product A will have a minor bump. Product B a patch. Product C a major version. Product D isn't impacted at all."

nx affected won't work in my case because it doesn't take into account cache inputs. nx affected would tell me, "naw, that project isn't impacted", and then I'd run the build and a change would get deployed.

nx cache won't help me answer this question, either, because it can only say, "yep, the deploy _for this branch_ has not changed" - I really need to answer the question "will this trigger a deploy _to prod_?"

I don't expect nx to deal with my unique situation OOTB, but I'd love to have a few more tools in the toolbox.
